### PR TITLE
Remove the direct dependency on Apache Commons. Issue - #7658

### LINKED ISSUE #7658: Remove the direct dependency on Apache Commons
### Feature Request / Improvement

See https://github.com/apache/iceberg/pull/5024/files#r896336657

### Query engine

None

### PR DESCRIPTION
**Pull request by Neuw84**
The followed approach is to use native Java 8 code when the alternative was clear or replicate/borrow when not following iceberg-core approach..

Flink and Spark projects declare Apache Commons as direct dependency and therefore on those subprojects the dependency is not removed.

Closes #7658

### CODE REVIEW

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 1]
```diff
@@ -662,15 +681,4 @@ private List<String> filterByExtension(String tablePath, String extension) {
         .filter(f -> f.endsWith(extension))
         .collect(Collectors.toList());
   }
-
-  private static String addRecordsToFile(Table table, String filename) throws IOException {
```
Comment (nastra) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:666]: why the change to move this from here to the top?

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetAvro.java [hunk 2]
```diff
@@ -41,6 +40,74 @@
 
 class ParquetAvro {
 
+  static final GenericData DEFAULT_MODEL =
```
Comment (nastra) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetAvro.java:43]: it's hard to follow which changes were exactly done in this file due to the large diff

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java [hunk 3]
```diff
@@ -21,7 +21,6 @@
 import java.io.IOException;
 import java.util.List;
 import java.util.Map;
-import org.apache.commons.lang3.ArrayUtils;
```
Comment (nastra) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java:24]: should compilation not fail here?

Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java:24]: I did not removed the libraries from Gradle for the subprojects ( Maven expert but not so much on Gradle), I have been searching but seems that the sub-projects were using the parent declared libraries when needed... 

Will fix this line on next re-push and if you can point me to where I can delete the libraries for subprojects will be great!

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 1 above)
Comment (Neuw84) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:666]: will review the changes and repush, maybe it was a thing of IntelliJ on reformatting.

**Comment by Neuw84** [comment]
Hi @nastra, 

Addressed your comments in the latest push, let me know your thoughts. 

One question, there was an error in the tests: 

`package StandardCharsets does not exist`

I am assuming that we are using Java 1.8 for the project no? Maybe there is some issue on the Github Actions config (although is weird because that package was introduced on 1.7).

File: gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import org.testcontainers.shaded.com.google.common.base.Preconditions;
+
+public class TestUtils {
```
Comment (nastra) [inline gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java:32]: `TestUtils` indicates that this is a class containing tests for a `Utils` class, so maybe rename this to something else. 
Also is this a 1:1 copy of the `TestUtils` in aws?

File: gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import org.testcontainers.shaded.com.google.common.base.Preconditions;
```
Comment (nastra) [inline gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java:30]: wrong import, should be `org.apache.iceberg.relocated.com.google....`

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java (hunk 3 above)
Comment (nastra) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java:24]: seems to be used here: https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java#L104

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java [hunk 6]
```diff
@@ -21,7 +21,6 @@
 import java.nio.charset.StandardCharsets;
 import java.sql.Date;
 import java.sql.Timestamp;
-import org.apache.commons.lang3.StringUtils;
```
Comment (nastra) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:24]: seems to be used here: https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java#L70

File: dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java [hunk 7]
```diff
@@ -20,18 +20,25 @@
 
 import com.emc.object.Range;
 import com.emc.object.s3.S3ObjectMetadata;
-import com.emc.object.shadow.org.apache.commons.codec.digest.DigestUtils;
```
Comment (nastra) [inline dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:23]: isn't that using a shaded version, meaning that we wouldn't have to change this file?

Comment (Neuw84) [reply dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:23]: I replaced the code for native Java 8 code, then that shaded version is not going to be needed.

Comment (Neuw84) [reply dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:23]: I have replaced with native Java code

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java (hunk 6 above)
Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:24]: Have changed that code to native Java 8 compliant code

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java (hunk 3 above)
Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/HiveVectorizedReader.java:24]: Have changed that code to native Java 8 compliant coe

File: gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java (hunk 4 above)
Comment (Neuw84) [reply gcp/src/test/java/org/apache/iceberg/gcp/gcs/TestUtils.java]: The files are not exactly the same and I do not like  ( but personal opinion here) the `test-jarscope` thing between projects.

Open to suggestions/guidance here!

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockLocalController.java [hunk 8]
```diff
@@ -276,4 +277,246 @@ public void setMessage(String message) {
...
+   *
+   * <p>This is useful to wrap ServletInputStreams. The ServletInputStream will block if you try to
+   * read content from it that isn't there, because it doesn't know whether the content hasn't
+   * arrived yet or whether the content has finished. So, one of these, initialized with the
+   * Content-length sent in the ServletInputStream's header, will stop it blocking, providing it's
+   * been sent with a correct content length.
+   *
+   * <p>This code is borrowed from `org.apache.commons:commons-io`
+   */
+  public class BoundedInputStream extends FilterInputStream {
```
[the remark this answers is not in the document]
Comment (Neuw84) [reply aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockLocalController.java:292]: As it is test code that is not going to be shipped I just though that this was good enough. Open to suggestions here ( it is not very clear why the code is using `BoundedInputStream` and I did not wanted to mess with it. 

Open to suggestions/guidance here.

**Comment by Neuw84** [comment]
Hi @nastra, 

Have addressed your comments and applied spotless everywhere. 

About the open questions let me know what your feelings are.

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestCommonsS3Utils.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class TestCommonsS3Utils {
```
Comment (nastra) [inline aws/src/test/java/org/apache/iceberg/aws/s3/TestCommonsS3Utils.java:32]: can we remove the `Test` prefix here and the `Test` suffix in `GCSCommonUtilsTest` as `Test` indicates that it contains tests. Why not just `IOUtils` like it was named before?

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java [hunk 10]
```diff
@@ -67,12 +66,13 @@ public static MapWork findMapWork(JobConf job) {
       LOG.debug("Initializing for input {}", inputName);
     }
     String prefixes = job.get(DagUtils.TEZ_MERGE_WORK_FILE_PREFIXES);
-    if (prefixes != null && !StringUtils.isBlank(prefixes)) {
-      // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null) {
+      if (prefixes.trim().isEmpty()) {
```
Comment (nastra) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:70]: is this correct?

File: dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java [hunk 11]
```diff
@@ -70,7 +72,15 @@ public InputStream createInputStream(Range range) {
...
-    metadata.setETag(DigestUtils.md5Hex(content));
+    MessageDigest md = null;
+    try {
+      md = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+    md.update(content);
+    byte[] digest = md.digest();
+    metadata.setETag(DatatypeConverter.printHexBinary(digest));
```
[the remark this answers is not in the document]
Comment (Neuw84) [reply dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:83]: Agreed, that is not a long term solution as for Java 9 onwards we are going to need another library. 

Fixing with native code.

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestCommonsS3Utils.java (hunk 9 above)
Comment (Neuw84) [reply aws/src/test/java/org/apache/iceberg/aws/s3/TestCommonsS3Utils.java:32]: Done!

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java (hunk 10 above)
Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:70]: Thanks for the catch. Fixed

File: dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java [hunk 12]
```diff
@@ -70,9 +71,29 @@ public InputStream createInputStream(Range range) {
...
+    }
+    md.update(content);
+    byte[] digest = md.digest();
+    metadata.setETag(bytesToHex(digest));
     metadata.setContentLength((long) content.length);
     metadata.setUserMetadata(userMetadata);
     return metadata;
   }
+
+  private static final char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();
```
Comment (nastra) [inline dell/src/test/java/org/apache/iceberg/dell/mock/ecs/ObjectData.java:88]: `static final` fields should generally be placed at the top of the java file

File: delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java [hunk 13]
```diff
@@ -532,13 +533,14 @@ private void addExternalDatafiles(
...
-      String decodedPath = new URLCodec().decode(path);
+      String decodedPath = URLDecoder.decode(path, StandardCharsets.UTF_8.name());
       if (dataFileUri.isAbsolute()) {
         return decodedPath;
       } else {
         return tableRoot + File.separator + decodedPath;
       }
-    } catch (DecoderException e) {
+    } catch (UnsupportedEncodingException e) {
+      // In Java 10+ this will not need try-catch
```
Comment (nastra) [inline delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:543]: is that comment relevant? I think we can probably remove it

File: gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSCommonUtils.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.gcp.gcs;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class GCSCommonUtils {
```
Comment (nastra) [inline gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSCommonUtils.java:26]: One file is called `CommonsS3...` whereas this one is called `GCSCommons...`. I'd suggest to put the `GCS` / `S3` part at the same position in the name

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java [hunk 15]
```diff
@@ -67,12 +66,13 @@ public static MapWork findMapWork(JobConf job) {
       LOG.debug("Initializing for input {}", inputName);
     }
     String prefixes = job.get(DagUtils.TEZ_MERGE_WORK_FILE_PREFIXES);
-    if (prefixes != null && !StringUtils.isBlank(prefixes)) {
-      // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null) {
+      if (!prefixes.trim().isEmpty()) {
```
Comment (nastra) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:70]: both ifs can be folded into a single if like in the previous version of the code

File: delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java (hunk 13 above)
Comment (Neuw84) [reply delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:543]: Yes, I suppose that the tooling will catch this when the project gets upgraded to a newer Java version. 

Removed

File: gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSCommonUtils.java (hunk 14 above)
Comment (Neuw84) [reply gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSCommonUtils.java:26]: done!

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java (hunk 15 above)
Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:70]: agreed, I have tested all the combinations just to be sure in a JShell 👍 . 

`import org.apache.commons.lang3.StringUtils;
String a=null;

if (a != null && !StringUtils.isBlank(a)) {
    System.out.println("nullllll check");
    }
if (a != null && !a.trim().isEmpty()) {
        System.out.println("null check native");
    }
a="";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("empty check ");
    }
if (a != null && !a.trim().isEmpty()) {
        System.out.println("empty check native");
    }
a=" ";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("blank check");
    }
if (a != null && !a.trim().isEmpty()) {
        System.out.println("blank check native");
    }
a="bob";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("things check");
    }
    if (a != null && !a.trim().isEmpty()) {
        System.out.println("things check native");
    }
a="  bob  ";
if (a != null && !StringUtils.isBlank(a)) {
        System.out.println("trim things check");
    }
    if (a != null && !a.trim().isEmpty()) {
        System.out.println("trim things check native");
    }`

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java [hunk 16]
```diff
@@ -67,10 +66,11 @@ public static MapWork findMapWork(JobConf job) {
...
     String prefixes = job.get(DagUtils.TEZ_MERGE_WORK_FILE_PREFIXES);
-    if (prefixes != null && !StringUtils.isBlank(prefixes)) {
-      // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null && !prefixes.trim().isEmpty()) {
+        // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
+        // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
+        return null;
+      }
```
Comment (nastra) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:73]: I think } is an extra one and needs to be removed?

**Review by nastra — COMMENTED** [review]
please run `./gradlew spotlessApply` to fix formatting

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java (hunk 16 above)
Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:73]: yep you are right, seems to be weird because yesterday I compiled everything.

File: hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java [hunk 17]
```diff
@@ -67,10 +66,10 @@ public static MapWork findMapWork(JobConf job) {
       LOG.debug("Initializing for input {}", inputName);
     }
     String prefixes = job.get(DagUtils.TEZ_MERGE_WORK_FILE_PREFIXES);
-    if (prefixes != null && !StringUtils.isBlank(prefixes)) {
-      // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
-      // So, we don't use the below code that would get the correct MapWork. See HIVE-16985.
-      return null;
+    if (prefixes != null && !prefixes.trim().isEmpty()) {
+        // Currently SMB is broken, so we cannot check if it's  compatible with IO elevator.
```
Comment (nastra) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java:70]: formatting seems to be off?

Comment (Neuw84) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/vector/CompatibilityHiveVectorUtils.java]: yes, fixed on latest commit. 

If you can retry the CI/CD pipeline it will be great.

Thanks!

**Review by nastra — COMMENTED** [review]
changes LGTM, but I think we also need to update `TestS3InputStream` / `TestS3FileIOIntegration` / `ManifestReadBenchmark`

**Comment by Neuw84** [comment]
Done with the required changes. 

For `ManifestReadBenchmark` I have also changed the `Files.createTemp` from Guava as it collides with `java.nio.File.Files` and the change was small.

**Review by nastra — APPROVED** [review]

**Review by Fokko — APPROVED** [review]

**Comment by Neuw84** [comment]
Hi @nastra, @Fokko, 

I have merged the changes from this pull request ( some tests changed)

https://github.com/apache/iceberg/pull/7996

Let me know whether I need to change anything!

**Comment by nastra** [comment]
@Neuw84 could you please rebase against latest master and squash all commits together?

**Comment by Neuw84** [comment]
> @Neuw84 could you please rebase against latest master and squash all commits together?

Done @nastra!
